File: examples/scala/src/main/scala/com/xebia/functional/xef/scala/conversation/contexts/DivergentTasks.scala
package com.xebia.functional.xef.scala.auto.contexts

import com.xebia.functional.xef.conversation.llm.openai.OpenAI
import com.xebia.functional.xef.prompt.Prompt
import com.xebia.functional.xef.reasoning.serpapi.Search
import com.xebia.functional.xef.scala.auto.*
import io.circe.Decoder

private final case class NumberOfMedicalNeedlesInWorld(numberOfNeedles: Long) derives SerialDescriptor, Decoder

@main def runDivergentTasks: Unit =
  conversation {
    val search = Search(OpenAI.FromEnvironment.DEFAULT_CHAT, summon[ScalaConversation], 3)
    addContext(search.search("Estimate amount of medical needles in the world").get())
    val needlesInWorld = prompt[NumberOfMedicalNeedlesInWorld](Prompt("Provide the number of medical needles in the world as an integer number"))
    println(s"Needles in world: ${needlesInWorld.numberOfNeedles}")
  }

File: examples/src/main/kotlin/com/xebia/functional/xef/conversation/contexts/DivergentTasks.kt
package com.xebia.functional.xef.conversation.contexts

import ai.xef.openai.StandardModel
import com.xebia.functional.openai.models.CreateChatCompletionRequestModel
import com.xebia.functional.xef.conversation.Conversation
import com.xebia.functional.xef.prompt.Prompt
import com.xebia.functional.xef.prompt.templates.user
import com.xebia.functional.xef.reasoning.serpapi.Search
import kotlinx.serialization.Serializable

@Serializable data class NumberOfMedicalNeedlesInWorld(val numberOfNeedles: Long)

suspend fun main() {
  Conversation {
    val model = StandardModel(CreateChatCompletionRequestModel.gpt_3_5_turbo_16k_0613)
    val search = Search(model = model, scope = this)
    addContext(search("Estimate amount of medical needles in the world"))
    val needlesInWorld: NumberOfMedicalNeedlesInWorld =
      prompt(Prompt(model) { +user("Provide the number of medical needles in the world") })
    println("Needles in world: ${needlesInWorld.numberOfNeedles}")
  }
}
